File: src/preprint_af/reasoners/latex.py
from __future__ import annotations

from agentfield import AgentRouter

from . import helpers
from .models import CompileReport

router = AgentRouter(prefix="latex", tags=["latex"])

MAX_ATTEMPTS = 3


def _repair_prompt(excerpt: str) -> str:
    return f"""\
LaTeX compilation failed. Fix ONLY LaTeX/compilation errors: undefined control sequences, missing
packages (prefer removing the dependency), math-mode errors, unescaped %, &, #, _ in text, missing
\\includegraphics files (replace with \\todobox{{Figure <name> pending}} if the pdf is genuinely
absent), bib issues. Do NOT change scientific content, numbers, citations, prose wording, or
section structure. Error log: {excerpt}"""


@router.reasoner()
async def compile_paper(workspace: dict, model: str | None = None) -> CompileReport:
    """P4: compile the paper with latexmk, repairing LaTeX errors between attempts (up to 3)."""
    root = workspace["root"]
    paper_dir = workspace["paper_dir"]

    last_excerpt = ""
    for attempt in range(1, MAX_ATTEMPTS + 1):
        ok, pdf_path, excerpt = helpers.run_latexmk(paper_dir)
        if ok:
            print(f"[latex] compile succeeded on attempt {attempt}: {pdf_path}")
            helpers.git_snapshot(root, f"P4 compile success (attempt {attempt})")
            return CompileReport(success=True, attempts=attempt, pdf_path=pdf_path)

        last_excerpt = excerpt
        print(f"[latex] compile attempt {attempt} failed")
        if attempt >= MAX_ATTEMPTS:
            break

        # One repair pass targeting only LaTeX/compilation errors.
        result = await router.harness(
            _repair_prompt(excerpt),
            provider="opencode",
            model=helpers.opencode_model(model),
            cwd=paper_dir,
            project_dir=root,
        )
        if result.is_error:
            print(f"[latex] repair pass (after attempt {attempt}) errored: {result.error_message}")

    print(f"[latex] compile failed after {MAX_ATTEMPTS} attempts")
    return CompileReport(success=False, attempts=MAX_ATTEMPTS, error_excerpt=last_excerpt)

File: go/internal/pipeline/latex.go
package pipeline

import (
	"context"
	"fmt"

	"github.com/Agent-Field/agentfield/sdk/go/harness"
	"github.com/Agent-Field/preprint-af/go/internal/prompts"
)

const maxLatexAttempts = 3

type CompilePaperInput struct {
	Workspace Workspace `json:"workspace"`
	Model     *string   `json:"model"`
}

func (s *Service) CompilePaper(ctx context.Context, in CompilePaperInput) (any, error) {
	last := ""
	for attempt := 1; attempt <= maxLatexAttempts; attempt++ {
		ok, pdf, excerpt := RunLatexmk(in.Workspace.PaperDir)
		if ok {
			GitSnapshot(in.Workspace.Root, fmt.Sprintf("P4 compile success (attempt %d)", attempt))
			return CompileReport{Success: true, Attempts: attempt, PDFPath: pdf}, nil
		}
		last = excerpt
		if attempt == maxLatexAttempts {
			break
		}
		if err := s.acquireHarness(ctx); err != nil {
			return nil, err
		}
		hr, err := s.App.Harness(ctx, prompts.LatexRepairPrompt(excerpt), nil, nil, harness.Options{Provider: "opencode", Model: OpenCodeModel(stringValue(in.Model)), Cwd: in.Workspace.PaperDir, ProjectDir: in.Workspace.Root})
		s.releaseHarness()
		if err != nil {
			fmt.Printf("[latex] repair pass errored: %v\n", err)
		} else if hr != nil && hr.IsError {
			fmt.Printf("[latex] repair pass errored: %s\n", hr.ErrorMessage)
		}
	}
	return CompileReport{Success: false, Attempts: maxLatexAttempts, ErrorExcerpt: last}, nil
}
